# Lets get Visual
We saw in the last notebook how to extract numerical information from our data - like the survival rates for men and women. We'll see in this notebook how we can display that kind of information visually. We'll use some techiniques we saw in the gapminder data, and learn some new ones as well

### Imports
Always remember to start your notebook by importing any libraries you will need. In this case we need pandas and plotly express, and I'll load colab's data_table extension so our data is a little prettier

In [0]:
import pandas as pd
import plotly.express as px

%load_ext google.colab.data_table

### Data
We always need to load our data as well. In this case we will be looking at AIDS data from around the world. 

Our data is usually in csv format, so we will usually load our data with `pd.read_csv`, but if our data is in another format we might need a different pandas function.

We pass the location and name of our file to `pd.read_csv` as a string. In this case our data is found at 
```
https://think.cs.vt.edu/corgis/datasets/csv/energy/energy.csv
```
So our call is:
```
pd.read_csv('https://think.cs.vt.edu/corgis/datasets/csv/energy/energy.csv')
```
Finally, we always need to save it to a variable so we can work with the data once it is loaded.

In [0]:
energy=pd.read_csv('https://think.cs.vt.edu/corgis/datasets/csv/energy/energy.csv')

We now have our data loaded and saved to our variable `aids`. to see our data we just type in aids on its own.

In [4]:
energy

State  ...  Price.Transportation.Natural Gas
0                     Utah  ...                              0.00
1                  Vermont  ...                              0.00
2             Rhode Island  ...                              0.00
3                 Michigan  ...                              0.00
4                 Missouri  ...                              0.00
...                    ...  ...                               ...
2800             Louisiana  ...                              8.08
2801          North Dakota  ...                              9.27
2802              Nebraska  ...                             16.23
2803  District of Columbia  ...                             11.75
2804                Alaska  ...                              8.28

[2805 rows x 85 columns]

We can see that our data has 2,805 rows and 85 columns. Each row represents the data for a particular state in a particular year. The columns represent a great deal of different pieces of information regarding energy sources and consumption. Read the table at https://think.cs.vt.edu/corgis/csv/energy/ to see what each column represents.

## Visualizations

We now need to decide how we are going to visualize our data. The kind of graph you want will depend on what you are trying to display. Plotly has a large number of graphs we can make, but the most common are:
- **Line Graph:** We usually use line graphs when we are showing the change in something over time. We use time on our x-axis, and the value we are looking at on our y-axis. Sometimes we use non-time variables on our x-axis, but it is less common.
- **Scatter Plot** We use scatter plots to show the relationship between two variables, one on the x-axis and one on the y-axis
- **Bar Chart** We use a bar chart to show the amount (how much stuff there is)of a variable in diffferent categories.  The categories may be time based, or they may be something else entirely
- **Pie Chart** We use a pie chart to show how a "whole" is broken up into parts. For instance, we might show how the whole population of Sky Islands is broken up into freshman, sophmores, juniors, seniors, and staff.

Let's see how we might analyze the `energy` data with each of these graph types.



### Line Graph
A natural question might be to find out how the consumption of a particular energy source has changed over time. To simplify our analysis, let's just look at Arizona.

In [0]:
# I'm just going to look at Arizona, so lets make an Arizona sub-frame
az_mask=energy['State']=='Arizona'
az_energy=energy[az_mask]

Now let's make a line graph showing how the consumption of natural gas as a source of energy for electric power has changed in Arizona:

In [6]:
px.line(az_energy,x='Year',y='Consumption.Electric Power.Natural Gas')

We can see that the consumption was fairly low until around 2000, when consumption began to rise quickly until consumption began to fall again. 
#### Improving it
Before we move on to our next graph, let's learn how we can make this one a bit better. When we're making a graph to show information, we always want a title to describe the graph, and our y axis label is unclear.

Luckily, Plotly makes it easy to give a graph a title, but it is slightly trickier to change the label. For the title we just add  `title='Our Title'` to our function call. 

For our y-axis label, or any other time we want to change the label of values from a column we pass a dictionary (we haven't covered these yet, so just copy my syntax) that has the form `{'name of column':'new label'}` we can rename as many columns this way as we want, we just put in a comma to separate them :

`{'name of coumn1':'new label1','name of column2':'new label2'}`

In [8]:
#making our graph and saving it to our variable, fig
px.line(az_energy, x='Year',y='Consumption.Electric Power.Natural Gas',
            # changing our title
            title='Natural Gas Consumption for electric power in Arizona',
            # changing our y label (which has data from the column Data.HIV Prevalence.Adults)
            labels={'Consumption.Electric Power.Natural Gas':'Natural gas consumption in BTUs'})



In [9]:
#making our graph and saving it to our variable, fig
px.line(az_energy, x='Year',y='Consumption.Electric Power.Natural Gas',
            # changing our title
            title='Natural Gas Consumption for electric power in Arizona',
            # changing our y label (which has data from the column Data.HIV Prevalence.Adults)
            labels={'Consumption.Electric Power.Natural Gas':'Natural gas consumption in BTUs', 'Year':'Hello, I am your x-label'})

### Scatter plot

We use a scatter plot when we want to see correlations between two variables. Let's compare electrical power consumption of coal and natural gas to see if there's a correlation. 




In [19]:
px.scatter(az_energy,x='Price.Electric Power.Coal',y='Consumption.Electric Power.Coal')

We see that there's a strong positive correlation - as price goes up, consumption goes up. This is unusual -- normally as prices go up people buy less, not more. Why do we see the opposite in this case?

#### Improving it
Let's add a title again, and change the labels to more readable labels (that also include our units). Notice that when we change our x and y labels, our data's labels also change when we hover over a point. 

Let's also add `'hover_name='Year'` so that the year will be displayed when we hover over a data point, perhaps we can understand our data better with this additional information.

In [25]:
px.scatter(az_energy,x='Price.Electric Power.Coal',y='Consumption.Electric Power.Coal',
            title='Arizona Coal consumption vs. Coal Price',
            labels={'Consumption.Electric Power.Coal': 'Coal Consumption (billion BTUs)',
                    'Price.Electric Power.Coal':'Cost of coal (Dollars/million BTU)'},
            hover_name='Year')

It seems like the older datapoints are in the lower left corner, and the more recent datapoints are in the upper right. If we assign color based on year, we might be able to see if this pattern holds.

In [27]:
px.scatter(az_energy,x='Price.Electric Power.Coal',y='Consumption.Electric Power.Coal',
            title='Arizona Coal consumption vs. Coal Price',
            labels={'Consumption.Electric Power.Coal': 'Coal Consumption (billion BTUs)',
                    'Price.Electric Power.Coal':'Cost of coal (Dollars/million BTU)'},
            hover_name='Year', color='Year')

That looks great! and explains some anomalies in our data.

### Bar Chart
Bar charts are good at displaying the quantity of something in different categories. In order to illustrate this example and the pie chart example, let's switch from looking at one state to one time:

In [0]:
mask_2000=energy['Year']==2000
#sorting by values so our states are in alphabetical order when we graph them
energy2000=energy[mask_2000].sort_values('State')

Now let's make a bar chart demonstrating coal consumption for electric power in each state:

In [36]:
px.bar(energy2000, x='State',y='Consumption.Electric Power.Coal',
        title='Coal Consumption for electric power by state in 2000',
        labels={'Consumption.Electric Power.Coal': 'Coal Consumption (billion BTUs)'})

We can see that the biggest consumer in 2000 was Texas, followed by Ohio, Indiana, and Pensylvania.

### Pie Chart
Pie Charts are good at showing the parts of a whole. In this case let's look at how total coal consumption for electric power is split up among the states. We'll look at the year 2000 again.



In [39]:
px.pie(energy2000, values='Consumption.Electric Power.Coal', names='State',
       title='Coal consumption in 2000 divided by state')

et voila

There are numerous other features of the graph you can change - the fonts, the color pallettes, 
the background color the tick marks and axes. Almost anything really. If you want to get fancy, go check out plotly's documentation at https://plot.ly/python/